### PR TITLE
Implement exploration gestures

### LINKED ISSUE #35: Implement exploration gestures
- [x] Next element
- [x] Previous element
- [x] Filter focusable elements

### LINKED ISSUE #36: Implement most important gestures
- [x] One-finger gesture support
- [x] Two-finger gesture support

### LINKED ISSUE #80: Improve Node Scan DSL
Improving node scan DSL.
- [x] Simplify stop scan function.
- [x] Simplify recursion of the scan descendants function.

### PR DESCRIPTION
**Pull request by Irineu333**
### Description

The exploration gestures allow you to navigate the screen through simple swipe gestures, in the case of Speak Touch, right to move forward and left to go back.

### Improvements
- Create DSL for nodes scan.
- Create gesture interceptor.
- Create focus controller.
- Implement move focus to next
- Implement move focus to previous

### Others
- Merge #36

### CODE REVIEW

**Review by PatrykMis — APPROVED** [review]
Surprised how great it is working, I can clearly say that swiping is much more stable than current open-source implementation of TalkBack.
One thing that should be done, I think (but for seperat issue) is automatic scrolling: for now, the current implementation changes focus only for visible on-screen things (I mean lists or grids) but does not scroll them if more elements are available. Scrolling should be done manually with two fingers up or down. but in fact it's already solid starting point and now it makes able to start using SpeakTouch on a daily basis as it's much faster than any currently known implementation.

**Comment by Irineu333** [comment]
@PatrykMis I will investigate this issue you raised.

**Comment by Irineu333** [comment]
@PatrykMis Very well observed, it really only works for visible elements on the screen. I will study how to solve this!

**Comment by PatrykMis** [comment]
as I tried to study, TalkBack does it by swiping the screen up/down/left/right if needed - that's why it needs `canPerformGestures`. However this approach is very slow, so would be worth investigating if there are no faster, more reliable solutions.

**Comment by Irineu333** [comment]
@PatrykMis  I got a working implementation of auto scroll. I think it can be improved, but for now I'm out of ideas.

File: app/src/main/java/com/neo/speaktouch/utils/extension/NodeScan.kt
Comment (Irineu333) [inline app/src/main/java/com/neo/speaktouch/utils/extension/NodeScan.kt:1]: Wow, I'm very proud of this solution.

**Comment by PatrykMis** [comment]
Wow, shocked how fast it is working comparing to TalkBack, amazing job!

**Comment by PatrykMis** [comment]
On Android 14, especially in Settings, sometimes it skipps first invisible item, right after screen has been scrolled, so scrolling is quite not stable yet. Unfortunately can't reproduce it each time so these things can be hard to debug. Scrolling is also not as fast as e.g. on iOS - maybe why in AndroidX Core v1.12.0 [AccessibilityNodeInfoCompat.AccessibilityActionCompat#ACTION_SCROLL_IN_DIRECTION](https://developer.android.com/reference/kotlin/androidx/core/view/accessibility/AccessibilityNodeInfoCompat.AccessibilityActionCompat#ACTION_SCROLL_IN_DIRECTION()) has been introduced. Unfortunately, I could not find any examples how to use it, though it seams a promising feature.

**Comment by Irineu333** [comment]
This bug was more frequent in the first attempts. I thought I had finally fixed it, but it persists. It's probably executing the callback twice, I'll fix it. Great job not letting that slip by!

**Comment by Irineu333** [comment]
Regarding ACTION_SCROLL_IN_DIRECTION, there are situations where it can be useful, for example when we can access something that is totally or partially off the screen, we can choose to bring just that one on the screen instead of scrolling the entire screen. But implementing this correctly can be challenging.

One that is completely off-screen, invisibleForUser can be used, but one that is partially off-screen, I have no idea.

**Comment by Irineu333** [comment]
@PatrykMis I believe it has been fixed now, in my tests it didn't skip a single time.

**Comment by Irineu333** [comment]
GPT liked our implementation: https://chat.openai.com/share/4f64552b-8922-421d-b212-491dfe21a85b

**Comment by Irineu333** [comment]
> as I tried to study, TalkBack does it by swiping the screen up/down/left/right if needed - that's why it needs `canPerformGestures`. However this approach is very slow, so would be worth investigating if there are no faster, more reliable solutions.

From what I understand, the `canPerformGestures` attribute is related to the fingerprint sensor

**Comment by Irineu333** [comment]
> On Android 14, especially in Settings, sometimes it skipps first invisible item, right after screen has been scrolled, so scrolling is quite not stable yet. Unfortunately can't reproduce it each time so these things can be hard to debug. Scrolling is also not as fast as e.g. on iOS - maybe why in AndroidX Core v1.12.0 [AccessibilityNodeInfoCompat.AccessibilityActionCompat#ACTION_SCROLL_IN_DIRECTION](https://developer.android.com/reference/kotlin/androidx/core/view/accessibility/AccessibilityNodeInfoCompat.AccessibilityActionCompat#ACTION_SCROLL_IN_DIRECTION()) has been introduced. Unfortunately, I could not find any examples how to use it, though it seams a promising feature.

It seems quite interesting, but I found that not every node supports it. There are a number of useful actions that are not widely supported, such as [AccessibilityNodeInfoCompat.AccessibilityActionCompat.ACTION_SCROLL_TO_POSITION](https://developer.android.com/reference/kotlin/androidx/core/view/accessibility/AccessibilityNodeInfoCompat.AccessibilityActionCompat#ACTION_SCROLL_TO_POSITION()). We can include it in a separate task, but it needs to be well tested.

**Review by PatrykMis — APPROVED** [review]

**Pull request merged**